ggml/src/ggml-cpu/llamafile/sgemm.cpp

```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
       } \
    } \
 
+#define COMPUTE(ACC, c_idx, s_idx) \
+    __builtin_mma_disassemble_acc(vec_C, ACC); \
+    for (int i = 0; i < 4; i++) { \
+        CA[i] = vec_splats((float)(((double)comparray[c_idx+i]) * -128.0)); \
+        res[i] = vec_add(vec_ctf(vec_C[i], 0), CA[i]); \
+        fin_res[s_idx+i] = vec_madd(res[i], vs[s_idx+i], fin_res[s_idx+i]); \
+    } \
+
+#define SAVE_RES(ii, jj, RM, RN, idx) \
+    for (int I = 0; I < RM; I++) { \
+        for (int J = 0; J < RN; J++) { \
+            *((float*)(C+ii+((jj+J)*ldc)+I)) = *((float*)&fin_res[idx+I]+J); \
+        } \
+    } \
+
+#define PERMUTE_VEC(IR1, IR2, IR3, IR4, vecOffset) \
+    t1 = vec_perm(IR1, IR2, swiz1); \
+    t2 = vec_perm(IR1, IR2, swiz2); \
+    t3 = vec_perm(IR3, IR4, swiz1); \
+    t4 = vec_perm(IR3, IR4, swiz2); \
+    t5 = vec_perm(t1, t3, swiz3); \
+    t6 = vec_perm(t1, t3, swiz4); \
+    t7 = vec_perm(t2, t4, swiz3); \
+    t8 = vec_perm(t2, t4, swiz4); \
+    if (flip == true) { \
+        t5 = vec_xor(t5, xor_vector); \
+        t6 = vec_xor(t6, xor_vector); \
+        t7 = vec_xor(t7, xor_vector); \
+        t8 = vec_xor(t8, xor_vector); \
+    } \
+    vec_xst(t5, 0, vecOffset); \
+    vec_xst(t6, 0, vecOffset+16); \
+    vec_xst(t7, 0, vecOffset+32); \
+    vec_xst(t8, 0, vecOffset+48); \
+
+template <typename TA, typename TB, typename TC>
+class tinyBLAS_Q0_PPC {
+  public:
+    tinyBLAS_Q0_PPC(int64_t k,
+                const TA *A, int64_t lda,
+                const TB *B, int64_t ldb,
+                TC *C, int64_t ldc,
+                int ith, int nth)
+        : A(A), B(B), C(C), k(k), lda(lda), ldb(ldb), ldc(ldc), ith(ith), nth(nth) {
+    }
+
+    void matmul(int64_t m, int64_t n) {
+        mnpack(0, m, 0, n);
+    }
+
+  private:
+    void (tinyBLAS_Q0_PPC::*kernel)(int64_t, int64_t);
```

Comment:
Did not see it last time, dangerous, it is not thread safe.

Reply:
Removed.